shared/src/balancer/event_handler.rs

```diff
@@ -235,75 +246,125 @@ impl PoolRegistry {
-    ) -> Result<Vec<(EventIndex, PoolCreated)>> {
-        contract_events
-            .into_iter()
-            .map(|EthContractEvent { data, meta }| {
-                let meta = match meta {
-                    Some(meta) => meta,
-                    None => return Err(anyhow!("event without metadata")),
-                };
-                match data {
-                    ContractEvent::PoolCreated(event) => convert_pool_created(&event, &meta),
-                }
-            })
-            .collect::<Result<Vec<_>>>()
+        token_pairs: HashSet<TokenPair>,
+    ) -> Vec<RegisteredWeightedPool> {
+        let mut pool_set_1 = self
+            .weighted_pool_updater
+            .lock()
+            .await
+            .store
+            .pools_containing_token_pairs(token_pairs.clone());
+        let pool_set_2 = self
+            .two_token_pool_updater
+            .lock()
+            .await
+            .store
+            .pools_containing_token_pairs(token_pairs);
+        pool_set_1.extend(pool_set_2);
+        pool_set_1
     }
 }
 
-pub struct BalancerEventUpdater(
-    Mutex<EventHandler<DynWeb3, BalancerV2WeightedPoolFactoryContract, PoolRegistry>>,
-);
-
-impl BalancerEventUpdater {
-    pub async fn new(contract: BalancerV2WeightedPoolFactory, pools: PoolRegistry) -> Result<Self> {
-        let deployment_block = match contract.deployment_information() {
-            Some(DeploymentInformation::BlockNumber(block_number)) => Some(block_number),
-            Some(DeploymentInformation::TransactionHash(hash)) => Some(
-                contract
-                    .raw_instance()
-                    .web3()
-                    .block_number_from_tx_hash(hash)
-                    .await?,
-            ),
-            None => None,
-        };
-        Ok(Self(Mutex::new(EventHandler::new(
-            contract.raw_instance().web3(),
-            BalancerV2WeightedPoolFactoryContract(contract),
-            pools,
-            deployment_block,
-        ))))
+impl BalancerPoolRegistry {
+    pub async fn new(web3: Web3, token_info_fetcher: Arc<dyn TokenInfoFetching>) -> Result<Self> {
+        let weighted_pool_factory = BalancerV2WeightedPoolFactory::deployed(&web3).await?;
+        let two_token_pool_factory = BalancerV2WeightedPool2TokensFactory::deployed(&web3).await?;
+        let deployment_block_weighted_pool =
+            get_deployment_block(weighted_pool_factory.deployment_information(), &web3).await;
+        let deployment_block_two_token_pool =
+            get_deployment_block(two_token_pool_factory.deployment_information(), &web3).await;
+        let weighted_pool_updater = Mutex::new(EventHandler::new(
+            web3.clone(),
+            BalancerV2WeightedPoolFactoryContract(weighted_pool_factory),
+            PoolStorage::new(Box::new(PoolDataFetcher {
+                web3: web3.clone(),
+                token_info_fetcher: token_info_fetcher.clone(),
```

Comment:
Is cloning the correct way to pass the token info fetcher into both event handlers? @fleupold.

Reply:
It is behind an `Arc`, so cloning is a very cheap operation (it just increases the reference count - i.e. the number of "owners" to the inner data).

It is the correct way of sharing an `Arc`.

Reply:
Cool, now I am just waiting on @fleupold's final thoughts and I would like to ship-it.